File: HaskellCocoa/Cocoa/AppKit/HNSView.hs


{-# LANGUAGE ForeignFunctionInterface #-}



module Cocoa.AppKit.HNSView
(
    HNSView
,   HNSViewObj(..)
,   nsViewCreate
,   nsView_addSubview
,   nsView_removeFromSuperview
,   nsView_setFrame
,   nsView_frame
,   nsView_setNeedsDisplay
,   nsView_inLiveResize
,   nsView_convertPointFromView
,   nsView_convertPointFromWindow
)
where


import Foreign
import Cocoa.Runtime.HNSObject
import Cocoa.Foundation.HNSGeometry



class (HNSObject a) => HNSView a where

    nsView_addSubview :: (HNSView b) => Ptr a -> Ptr b -> IO ()
    nsView_addSubview = hns_view_addSubview

    nsView_removeFromSuperview :: Ptr a -> IO ()
    nsView_removeFromSuperview = hns_view_removeFromSuperview

    nsView_setFrame :: Ptr a -> HNSRect -> IO ()
    nsView_setFrame view (HNSRect x y width height) = hns_view_setFrame view x y width height

    nsView_frame :: Ptr a -> IO HNSRect
    nsView_frame view = do
                            rectP <- hns_view_frame view
                            peek rectP

    nsView_setNeedsDisplay :: Ptr a -> Bool -> IO ()
    nsView_setNeedsDisplay = hns_view_setNeedsDisplay

    nsView_inLiveResize :: Ptr a -> IO Bool
    nsView_inLiveResize = hns_view_inLiveResize

    nsView_convertPointFromView :: Ptr a -> HNSPoint -> Ptr a -> IO HNSPoint
    nsView_convertPointFromView view (HNSPoint x y) fromView =
                        do
                            pointPtr <- hns_view_convertPointFromView view x y fromView
                            peek pointPtr

    nsView_convertPointFromWindow :: Ptr a -> HNSPoint -> IO HNSPoint
    nsView_convertPointFromWindow view (HNSPoint x y) =
                        do
                            pointPtr <- hns_view_convertPointFromWindow view x y
                            peek pointPtr





data HNSViewObj = HNSViewObj


instance HNSObject HNSViewObj where

instance HNSView HNSViewObj where


nsViewCreate :: IO (Ptr HNSViewObj)
nsViewCreate = hns_ViewCreate




foreign import ccall hns_ViewCreate :: IO (Ptr HNSViewObj)

foreign import ccall hns_view_addSubview :: Ptr a -> Ptr b -> IO ()
foreign import ccall hns_view_removeFromSuperview :: Ptr a -> IO ()

foreign import ccall hns_view_setFrame :: Ptr a -> Double -> Double -> Double -> Double -> IO ()
foreign import ccall hns_view_frame :: Ptr a -> IO (Ptr HNSRect)

foreign import ccall hns_view_convertPointFromView :: Ptr a -> Double -> Double -> Ptr a -> IO (Ptr HNSPoint)
foreign import ccall hns_view_convertPointFromWindow :: Ptr a -> Double -> Double -> IO (Ptr HNSPoint)
foreign import ccall hns_view_setNeedsDisplay :: Ptr a -> Bool -> IO ()
foreign import ccall hns_view_inLiveResize :: Ptr a -> IO Bool

File: HaskellCocoa/Cocoa/AppKit/HNSView.m
//
//  HNSObject.m
//  
//
//  Created by middleware on 2/13/16.
//
//

#import <AppKit/AppKit.h>

#import "HNSView.h"
#import "HNSObject.h"
#import "HNSGeometry.h"



HsPtr hns_ViewCreate(void)
{
    NSView* pView = [[NSView alloc] init];
    hns_retain(pView);
    return (__bridge HsPtr)pView;
}


void hns_view_addSubview(HsPtr view, HsPtr subview)
{
    NSView* pView = (__bridge NSView *)(view);
    NSView* pSubview = (__bridge NSView *)(subview);
    [pView addSubview:pSubview];
}



void hns_view_removeFromSuperview(HsPtr view)
{
    NSView* pView = (__bridge NSView *)(view);
    [pView removeFromSuperview];
}


void hns_view_setFrame(HsPtr view, double x, double y, double width, double height)
{
    NSView* pView = (__bridge NSView *)(view);
    NSRect frame = NSMakeRect(x, y, width, height);
    [pView setFrame:frame];
}


void hns_view_setNeedsDisplay(HsPtr view, HsBool need)
{
    NSView* pView = (__bridge NSView *)(view);
    [pView setNeedsDisplay:need];
}


HsBool hns_view_inLiveResize(HsPtr view)
{
    NSView* pView = (__bridge NSView *)(view);
    return pView.inLiveResize;
}



HsPtr hns_view_convertPointFromView(HsPtr view, HsDouble x, HsDouble y, HsPtr fromView)
{
    NSView* pView = (__bridge NSView *)(view);
    NSView* pFromView = (__bridge NSView *)(fromView);
    NSPoint point = NSMakePoint(x, y);
    
    static struct HsPoint result;
    NSPoint resPoint = [pView convertPoint:point fromView:pFromView];
    result.x = resPoint.x;
    result.y = resPoint.y;
    return &result;
}


HsPtr hns_view_frame(HsPtr view)
{
    static struct HsRect result;
    NSView* pView = (__bridge NSView *)(view);
    NSRect frame = pView.frame;
    result.x = frame.origin.x;
    result.y = frame.origin.y;
    result.w = frame.size.width;
    result.h = frame.size.height;
    return &result;
}
